web-common/src/WebSocket/Support.purs

```diff
@@ -0,0 +1,204 @@
+import Prim.TypeError (class Warn, Text)
+import Web.Event.EventTarget (addEventListener, eventListener)
+import Web.HTML as W
+import Web.HTML.Location as WL
+import Web.HTML.Window as WW
+import Web.Socket.Event.CloseEvent (CloseEvent)
+import Web.Socket.Event.CloseEvent as CloseEvent
+import Web.Socket.Event.EventTypes (onClose, onMessage, onOpen)
+import Web.Socket.Event.MessageEvent (MessageEvent)
+import Web.Socket.Event.MessageEvent as MessageEvent
+import Web.Socket.WebSocket (WebSocket)
+import Web.Socket.WebSocket as WS
+
+-- | A type that manages a queue of messages going-to and coming-from a websocket.
+-- |
+-- | You may ask, "Why not use the websocket directly?" and the answer
+-- | is so we can transparently manage network problems and reconnections
+-- | for you. Messages go into inbound and outbound queues, and if the
+-- | connection drops we hold the queue while we get it back up.
+-- |
+-- | See `runWebSocketManager` for an example of how to use this.
+newtype WebSocketManager i o
+  = WebSocketManager
+  { inbound :: Queue (FromSocket i)
+  , outbound :: Queue (ToSocket o)
+  , socket :: AVar WebSocket
+  }
+
+derive instance newtypeWebSocketManager :: Newtype (WebSocketManager i o) _
+
+managerReadInbound :: forall i o. WebSocketManager i o -> Aff (FromSocket i)
+managerReadInbound = unwrap >>> _.inbound >>> Queue.read
+
+managerWriteInbound :: forall i o. WebSocketManager i o -> FromSocket i -> Aff Unit
+managerWriteInbound = unwrap >>> _.inbound >>> Queue.write
+
+managerReadOutbound :: forall i o. WebSocketManager i o -> Aff (ToSocket o)
+managerReadOutbound = unwrap >>> _.outbound >>> Queue.read
+
+managerWriteOutbound :: forall i o. WebSocketManager i o -> ToSocket o -> Aff Unit
+managerWriteOutbound = unwrap >>> _.outbound >>> Queue.write
+
+mkWebSocketManager :: forall i o. Aff (WebSocketManager i o)
+mkWebSocketManager = do
+  inbound <- Queue.new
+  outbound <- Queue.new
+  socket <- liftEffect AVar.empty
+  pure $ WebSocketManager { inbound, outbound, socket }
+
+------------------------------------------------------------
+data FromSocket a
+  = WebSocketOpen
+  | ReceiveMessage (Either MultipleErrors a)
+  | WebSocketClosed CloseEvent
+
+data ToSocket a
+  = SendMessage a
+
+------------------------------------------------------------
+newtype URI
+  = URI String
+
+mkSocket ::
+  URI -> Effect WebSocket
+mkSocket (URI uri) = do
+  window <- W.window
+  location <- WW.location window
+  protocol <- WL.protocol location
+  hostname <- WL.hostname location
+  port <- WL.port location
+  let
+    wsProtocol = case protocol of
+      "https:" -> "wss"
+      _ -> "ws"
+
+    wsPath = wsProtocol <> "://" <> hostname <> ":" <> port <> uri
+  WS.create wsPath []
+
+------------------------------------------------------------
+-- | Take a resource that's wrapped in an AVar, then lock it, use it and unlock it.
+bracket :: forall m a. MonadEffect m => (a -> Effect Unit) -> AVar a -> m Unit
+bracket f resourceVar =
```

Comment:
should you pass in the error handler really?

Reply:
I mean the function `bracket` here has been defined in quite a general way, it seems to me that logging error handler that's provided as part of the implementation doesn't really fit. Isn't this function only used once? If so then you don't _need_ to make the success handler an argument but if you do (and I think it's nice) then IMO you should also make the error handler an argument

Reply:
Yup, fixed.